---
title: "scTrancriptomics in plants"
subtitle: "Tutorial 1"
## date: "23/04/2023"

output:
  xaringan::moon_reader:
    css: [default, metropolis, metropolis-fonts]
    seal: false
    nature:
      ratio: '16:9'
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---
class: left, middle
####Computer Practice 1

##Single cell Transcriptomics using R
##Seurat: from counts to clusters
####Victoria Mironova 
####Associate Professor, Department of Plant Systems Physiology
<img src="Figures/Theme2.png" width="100px" align = 'right'>

---
```{r include = FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, fig.retina = 3)
set.seed(100)
```

.pull-left[
#Course structure

- C1. From reads to counts
- C2. From counts to clusters
- C3. From clusters to marker genes
- C4. Developmental trajectories, developmental states.

The course materials can be found via [GitHub](https://github.com/VictoriaVMironova/ScTranscriptomics_in_plants)

]
---
#Step 0. Preparatory: install packages

First we will need to install packages and load the libraries into your environmen. </br>
1. Use the file `_packages.R` from [GitHub](https://github.com/VictoriaVMironova/ScTranscriptomics_in_plants) to install the packages needed in a course.</br>

2. Then load the libraries. Today we need only Seurat and ggplot2 packages:

```{r libraries, message = FALSE}
library(Seurat)
library(ggplot2)

set.seed(42)
```

---
#Step 0. Preparatory: data download

In the computer practice, we will use the single-cell dataset from Arabidopsis leaf:
.pull-left[
*Lopez-Anido CB, Vat√©n A, Smoot NK, Sharma N et al. Single-cell resolution of lineage trajectories in the Arabidopsis stomatal lineage and developing leaf. Dev Cell 2021 Apr 5;56(7):1043-1055.e4.*
]
.pull-right[
```{r Ara leaf, echo=FALSE, fig.cap="Graphical summary from (Lopez et al., 2021)", out.width = '80%'}
knitr::include_graphics("Figures/ScLeaf.jpg")
```
]
---
#Step 0. Data download

The data is stored on GEO database: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE167135
We will use only the sample GSM509788 (ATML1p 10X Genomics pool 1). </br>

You can start from the scratch, downloading *.tar* archive yourself and finding the data matrix there.</br>
Or you can use the data we already downloaded and stored for you in `Data/GSM5097888`.

---
# Step 1. Data loading into R

With 10xGenomics datasets you can use `Read10x` function to load the data.

```{r pressure}
#add the path to the folder with the matrix here:
path <- "Data/GSM5097888"
leaf.data<-Read10X(path, gene.column = 1)
```

--
```{r raw matrix}
head(leaf.data)
dim(leaf.data)
```

--
There are 31110 genes and 5021 cells in the dataset

---
# Step 2.0. Making Seurat object

```{r seurat object}
leaf.dataset <- CreateSeuratObject(counts = leaf.data, project = "leaf")
```
--
```{r in seurat object}
head(leaf.dataset)
dim(leaf.dataset)
```
this will require to elaborate on more
---
# Step 2.1. Testing the dataset quaility

```{r featurescatter}
FeatureScatter(leaf.dataset, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

---
#Step 2.1. Testing the dataset quaility

.pull-left[
```{r mt and ct}
leaf.dataset[["percent.mt"]] <- PercentageFeatureSet(leaf.dataset, pattern = "^ATM")
leaf.dataset[["percent.ct"]] <- PercentageFeatureSet(leaf.dataset, pattern = "^ATC")
```
The `[[ ]]` operator can add columns to the object metadata. 
Check the Seurat object meta.data: we've got additional variable percent.mt and percent.ct in there calculated for all the cells.
]
.pull-right[
```{r volcano}
VlnPlot(object = leaf.dataset, features = c("nFeature_RNA", "nCount_RNA","percent.ct","percent.mt"), ncol = 5)
```
]
---
#Step 2.2. Filter out bad quality cells

```{r filter cells}
dim(leaf.dataset)
leaf.dataset <- subset(leaf.dataset, subset = percent.mt <= 20 & percent.ct <= 20 & nCount_RNA >=1000)
dim(leaf.dataset)
```

Only one cell was filtered out, this is because the dataset has been already cleaned by the authors.
---
#Step 3. Data normalization 

Transformed data will be available in the SCT assay, which is set as the default after running sctransform.
During normalization, we can also remove confounding sources of variation, for example, mitochondrial mapping percentage

```{r sctTransfrom, eval = FALSE}
leaf.dataset <- SCTransform(leaf.dataset)
```
This step might take some time to run...</br>

Note that this single command replaces NormalizeData(), ScaleData(), and FindVariableFeatures() functions used in earlier Seurat versions.

---
#Step 4. Data clustering and visualization

These are more or less standard steps in the single cell data clustering.

```{r clustering, eval = FALSE}
leaf.dataset <- RunPCA(leaf.dataset,verbose = FALSE)
leaf.dataset <- RunUMAP(leaf.dataset, dims = 1:50, verbose = FALSE)

leaf.dataset <- FindNeighbors(leaf.dataset, dims = 1:50, verbose = FALSE)
leaf.dataset <- FindClusters(leaf.dataset, resolution = 1,verbose = FALSE)
```

```{r load RDS1, show = FALSE}
leaf.dataset <- readRDS('Data/leaf.dataset.rds')
```
---
#Step 4. Data clustering and visualization
.pull-left[
```{r DimPLot 1, fig.height = 6}
DimPlot(leaf.dataset, label = TRUE)
```
]
.pull-right[
```{r DimPlot2, fig.height = 6}
DimPlot(leaf.dataset, label = TRUE, pt.size = 1.5, label.size = 10) + NoLegend()
```
]
---
#Step 5. Saving the Seurat object

As steps 3-4 are machine-time consuming. It is better to save the Seurat object at this stage as .rds file.

```{r save RDS, eval = FALSE}
saveRDS(leaf.dataset,'Data/leaf.dataset.rds')
```

So you can call the Seurat object in your future analysis directly.

```{r load RDS2, eval = FALSE}
leaf.dataset <- readRDS('Data/leaf.dataset.rds')
```
---
#Step 6. Gene expression vizualization
FeaturePlot visualizes the features stored in Seurat object. It is ggplot2 object, so you can adjust the color, legend, axes, etc.
.pull-left[
```{r FeaturePlot standard, fig.height = 6}
FeaturePlot(leaf.dataset,features = 'AT1G77990')
```
]
.pull-right[
```{r featureplot adj, fig.height = 6}
FeaturePlot(leaf.dataset,features = 'AT1G77990',order = T, label = T,pt.size = 3)+scale_color_gradientn(colors = c('lightgray','yellow','red','darkred'))
```
]
---
#Step 6. Gene expression vizualization
You can build FeaturePlot it for a number of genes at a time
```{r featureplot combined}
FeaturePlot(leaf.dataset,features = c('AT1G12480', 'AT1G11850', 'AT2G05100', 'AT5G38410'),order = T,label = F,pt.size = 3)
```
---
#Step 6. Gene expression vizualization 

```{r doHeatmap, fig.height= 6}
marker_genes<- c('AT2G05100', 'AT5G01530', 'AT4G10340', 'AT1G67090', 'AT5G38410', 'AT3G54890', 'AT1G29920', 'AT3G08940', 'AT3G27690', 'AT5G38430', 'AT1G76570', 'AT4G21750', 'AT2G26330', 'AT3G45640', 'AT2G42840', 'AT5G53210', 'AT5G60880', 'AT2G26330', 'AT1G80080', 'AT1G34245', 'AT3G26744', 'AT1G12860', 'AT2G42840', 'AT5G53210', 'AT5G60880', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT1G80080', 'AT1G34245', 'AT1G04110', 'AT4G31805', 'AT3G26744', 'AT1G12860', 'AT3G45640', 'AT3G06120', 'AT5G53210', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT1G80080', 'AT1G34245', 'AT2G20875', 'AT1G04110', 'AT4G31805', 'AT3G26744', 'AT1G12860', 'AT3G45640', 'AT3G06120', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT2G20875', 'AT1G04110', 'AT4G31805', 'AT3G26744', 'AT1G12860', 'AT1G14350', 'AT3G45640', 'AT3G24140', 'AT3G06120', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT2G20875', 'AT1G04110', 'AT3G26744', 'AT1G12860', 'AT1G14350', 'AT3G45640', 'AT3G24140', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT1G04110', 'AT1G14350', 'AT5G46240', 'AT3G45640', 'AT3G17300', 'AT5G46240', 'AT2G26330', 'AT3G45640', 'AT4G12970', 'AT2G17950', 'AT2G27250', 'AT2G45190', 'AT2G27250', 'AT4G17710', 'AT1G55580', 'AT2G22850', 'AT3G28730', 'AT4G32880', 'AT5G16560')

DoHeatmap(leaf.dataset, features = marker_genes) + NoLegend()
```

---
#Step 7. Clusters annotation by an expert

.pull-left[
```{r DimPLot 3, fig.height = 6}
DimPlot(leaf.dataset, label = TRUE, pt.size = 1.5, label.size = 10) + NoLegend()
```
]
.pull-right[
```{r featureplot1, fig.height = 6}
DoHeatmap(leaf.dataset, features = marker_genes) + NoLegend()
```
]
---
#Step 7. Clusters annotation by an expert
.pull-left[
```{r DoHetamap clust, fig.height = 6}
DoHeatmap(leaf.dataset, features = marker_genes) + NoLegend()
```
]
.pull-right[
- AT2G05100 is LHCB2, LIGHT-HARVESTING CHLOROPHYLL B-BINDING 2
- AT5G01530, AT4G10340, AT3G54890, AT3G08940 are other subunits of LIGHT HARVESTING COMPLEX PHOTOSYSTEM II
- AT1G67090, AT5G38410, AT5G38430 encode Rubisco small subunits (RBCS).
...

--

What are the clusters 1 and 2 then?

]
---
#Step 7. Clusters reannotation

When you are ready to name a cluster you can use `RenameIdents` function. Here we will rename clusters 1 and 2 into 'Mesophyll' cell type.
```{r clusters annotation}
leaf.dataset <- RenameIdents(leaf.dataset,
                                                           `1`="Mesophyll",
                                                           `2`='Mesophyll')
leaf.dataset$CellTypes<-Idents(leaf.dataset)
```

---
#Step 7. Clusters reannotation
.pull-left[
```{r DimPLot 4, fig.height = 6}
DimPlot(leaf.dataset, label = TRUE, pt.size = 1.5, label.size = 10) + NoLegend()
```
]
.pull-right[
```{r reference data, fig.height = 5, ref.label='Ara leaf'}
```

]
---
#Your turn:

Explore gene expression for other known marker genes in Arabbidopsis leaf (find them using google).

Name other clusters in the dataset.


