---
title: "scTrancriptomics in plants"
subtitle: "Tutorial 2"
## date: "23/04/2023"

output:
  xaringan::moon_reader:
    css: [default, metropolis, metropolis-fonts]
    seal: false
    nature:
      ratio: '16:9'
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---
class: left, middle
####Computer Practice 2

##Single cell Transcriptomics using R
##Seurat: from counts to clusters
####Victoria Mironova 
####Associate Professor, Department of Plant Systems Physiology
<img src="Figures/Theme2.png" width="100px" align = 'right'>

---
class: middle, inverse
```{r include = FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, fig.retina = 3)
set.seed(100)
```

.pull-left[
#Course structure

- C1. From reads to counts
- C2. From counts to clusters
- C3. From clusters to marker genes
- C4. Subsetting and integrating the data.
- C5. Developmental trajectories, developmental states.

The course materials can be found via [GitHub](https://github.com/VictoriaVMironova/ScTranscriptomics_in_plants)

]
---
#Step 0. Data download

In the computer practice, we use the single-cell dataset from Arabidopsis leaf:
.pull-left[
*Lopez-Anido CB, Vat√©n A, Smoot NK, Sharma N et al. Single-cell resolution of lineage trajectories in the Arabidopsis stomatal lineage and developing leaf. Dev Cell 2021 Apr 5;56(7):1043-1055.e4.*
]
.pull-right[
```{r Ara leaf, echo=FALSE, fig.cap="Graphical summary from (Lopez et al., 2021)", out.width = '80%'}
knitr::include_graphics("Figures/ScLeaf.jpg")
```
]
---
#Step 0. Data download

The data is stored on GEO database: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE167135</br>
There are four datasets generated using 10xGenomics:
- **GSM509788 (ATML1p 10X Genomics pool 1)**.
- **GSM509789 (TMMp 10X Genomics replica 1)**,
- GSM509790 (TMMp 10X Genomics replica 2),
- GSM509791 (TMMp 10X Genomics replica 3).

--

This presentation uses GSM509788 (high quality data). </br>
We suggest you to use GSM509789 (lower quality data, smaller size). </br>

--
</br>
You can start from the scratch, downloading *.tar* archive yourself and finding the data matrix there.</br>
Or you can use the data we already downloaded and stored for you:</br>

GSM509788: in [Google Drive](https://drive.google.com/drive/folders/11hHUQ3nabTj230urqKnTXPCwoqWhZ8Zc?usp=sharing).

GSM509789: in [GitHub ](https://github.com/VictoriaVMironova/ScTranscriptomics_in_plants) repository `Data/GSM509789/`.

---
class: inverse, middle

#How this presentation works?
- All the course materials are uploaded to [GitHub](https://github.com/VictoriaVMironova/ScTranscriptomics_in_plants).
- You need to create a new project cloning the [GitHub](https://github.com/VictoriaVMironova/ScTranscriptomics_in_plants) repository. Or downloading the repository with all the course materials (including this tutorial) and opening *ScTranscriptomics_inPlants.Rproj* project from there. </br>
- This presentation is generated from .Rmd file which you can also find in the repository.
- The presentation visualizes the output of the code chunks that *CP2_from_counts_to_clusters.R* contains.
- We will go through the code line-by-line together. 

---
#Step 0. Preparatory: install packages

First we will need to install packages and load the libraries into your environment. </br>

1. Use the file `_packages.R` to install the packages needed in a course.</br>

2. open the script `CP2_from_counts_to_clusters.R` for today's practice.

3. Load the libraries using the script. Today we need only Seurat and tidyverse packages:

```{r libraries, message = FALSE}
library(Seurat)
library(tidyverse)

set.seed(42)
```

---
# Step 1. Data loading into R

With 10xGenomics datasets you can use `Read10x` function to load the data.

```{r data loading}
path <- "Data/GSM5097888"
leaf.counts<-Read10X(path, gene.column = 1)
```

--

In your case the path will be different.

--
```{r raw matrix}
dim(leaf.counts)
head(leaf.counts)
```

---
# Step 2.0. Making Seurat object

```{r seurat object}
leaf.dataset <- CreateSeuratObject(counts = leaf.counts, project = "leaf")
```
--
```{r in seurat object}
head(leaf.dataset)
dim(leaf.dataset)
```
--
How many cells there are in your case?
---
# Step 2.1. Testing the dataset quaility

```{r featurescatter}
FeatureScatter(leaf.dataset, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

---
#Step 2.1. Testing the dataset quaility

.pull-left[
```{r mt and ct}
leaf.dataset[["percent.mt"]] <- PercentageFeatureSet(leaf.dataset, pattern = "^ATM")
leaf.dataset[["percent.ct"]] <- PercentageFeatureSet(leaf.dataset, pattern = "^ATC")
```
The `[[ ]]` operator can add columns to the object metadata. 
Check the Seurat object meta.data: we've got additional variable percent.mt and percent.ct in there calculated for all the cells.
]
.pull-right[
```{r violine, fig.height= 6.5}
VlnPlot(object = leaf.dataset, features = c("nFeature_RNA", "nCount_RNA","percent.ct","percent.mt"), ncol = 5)
```
]
---
#Step 2.2. Filter out bad quality cells

```{r filter cells}
dim(leaf.dataset)
leaf.dataset <- subset(leaf.dataset, subset = percent.mt <= 20 & percent.ct <= 20 & nCount_RNA >=500)
dim(leaf.dataset)
```

Only one cell was filtered out, this is because the dataset has been already cleaned by the authors.</br>
--
What about your dataset?
---
#Step 3. Data normalization 

Transformed data will be available in the SCT assay, which is set as the default after running sctransform.

```{r sctTransfrom, eval = FALSE}
leaf.dataset <- SCTransform(leaf.dataset)
```
This step might take some time to run...</br>

Note that this single command replaces NormalizeData(), ScaleData(), and FindVariableFeatures() functions used in earlier Seurat versions.

---
#Step 4. Data clustering and visualization

These are more or less standard steps in the single cell data clustering.

```{r clustering, eval = FALSE}
leaf.dataset <- RunPCA(leaf.dataset,verbose = FALSE)
leaf.dataset <- RunUMAP(leaf.dataset, dims = 1:50, verbose = FALSE)

leaf.dataset <- FindNeighbors(leaf.dataset, dims = 1:50, verbose = FALSE)
leaf.dataset <- FindClusters(leaf.dataset, resolution = 1,verbose = FALSE)

```

```{r load RDS1, echo = FALSE}
leaf.dataset <- readRDS('Data/leaf.dataset.rds')
```
---
#Step 4. Explore the cell clusters

How many cells are in each cluster?
```{r clusters count}
table(Idents(leaf.dataset))
```
--
What proportion of cells are in each cluster?
```{r clusters prop}
prop.table(table(Idents(leaf.dataset)))
```
---
#Step 4. Data clustering and visualization
.pull-left[
```{r DimPLot 1, fig.height = 6}
DimPlot(leaf.dataset, label = TRUE)
```
]
.pull-right[
```{r DimPlot2, fig.height = 6}
DimPlot(leaf.dataset, label = TRUE, pt.size = 1.5, label.size = 10) + NoLegend()
```
]
---
#Step 5. Saving the Seurat object

As steps 3-4 are machine-time consuming. It is better to save the Seurat object at this stage as .rds file.

```{r save RDS, eval = FALSE}
saveRDS(leaf.dataset,'Data/leaf.dataset.rds')
```

So you can call the Seurat object in your future analysis directly.

```{r load RDS2, eval = FALSE}
leaf.dataset <- readRDS('Data/leaf.dataset.rds')
```
---
#Step 6. Gene expression vizualization

.pull-left[
```{r FeaturePlot standard, fig.height = 6}
FeaturePlot(leaf.dataset,features = 'AT3G24140')


```
]
.pull-right[
```{r featureplot adj, fig.height = 6}
FeaturePlot(leaf.dataset,features = 'AT3G24140',order = T, label = T,pt.size = 3)+
  scale_color_gradientn(colors = c('lightgray','yellow','red','darkred'))+
  ggtitle("FAMA")
```
]
---
#Step 6. Gene expression vizualization
You can build FeaturePlot for a number of genes at a time
```{r featureplot combined}
FeaturePlot(leaf.dataset,features = c('AT1G12480', 'AT1G11850', 'AT2G05100', 'AT5G38410'),order = T,label = F,pt.size = 3)
```
---
#Step 6. Gene expression vizualization 

```{r doHeatmap, fig.height= 6}
marker_genes<- c('AT3G24140', 'AT2G05100', 'AT5G01530', 'AT4G10340', 'AT1G67090', 'AT5G38410', 'AT3G54890', 'AT1G29920', 'AT3G08940', 'AT3G27690', 'AT5G38430', 'AT1G76570', 'AT4G21750', 'AT2G26330', 'AT3G45640', 'AT2G42840', 'AT5G53210', 'AT5G60880', 'AT2G26330', 'AT1G80080', 'AT1G34245', 'AT3G26744', 'AT1G12860', 'AT2G42840', 'AT5G53210', 'AT5G60880', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT1G80080', 'AT1G34245', 'AT1G04110', 'AT4G31805', 'AT3G26744', 'AT1G12860', 'AT3G45640', 'AT3G06120', 'AT5G53210', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT1G80080', 'AT1G34245', 'AT2G20875', 'AT1G04110', 'AT4G31805', 'AT3G26744', 'AT1G12860', 'AT3G45640', 'AT3G06120', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT2G20875', 'AT1G04110', 'AT4G31805', 'AT3G26744', 'AT1G12860', 'AT1G14350', 'AT3G45640', 'AT3G24140', 'AT3G06120', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT2G20875', 'AT1G04110', 'AT3G26744', 'AT1G12860', 'AT1G14350', 'AT3G45640', 'AT3G24140', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT1G04110', 'AT1G14350', 'AT5G46240', 'AT3G45640', 'AT3G17300', 'AT5G46240', 'AT2G26330', 'AT3G45640', 'AT4G12970', 'AT2G17950', 'AT2G27250', 'AT2G45190', 'AT2G27250', 'AT4G17710', 'AT1G55580', 'AT2G22850', 'AT3G28730', 'AT4G32880', 'AT5G16560')

DoHeatmap(leaf.dataset, features = marker_genes)
```
---
#Step 6. Adjusting heatmap
```{r doHeatmap adj, fig.height= 6}
DoHeatmap(leaf.dataset, features = marker_genes) + NoLegend()
    scale_y_discrete(labels = c("AT3G24140" = "FAMA")) +
    scale_fill_continuous(type = "viridis")
```

---
class: middle, inverse
#Your turn:

- Explore gene expression patterns for your dataset.

- Try to identify/name the clusters.


