<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>scTrancriptomics in plants</title>
    <meta charset="utf-8" />
    <script src="ComputerPractice3_from_clusters_to_genes_files/header-attrs-2.20/header-attrs.js"></script>
    <link href="ComputerPractice3_from_clusters_to_genes_files/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="ComputerPractice3_from_clusters_to_genes_files/remark-css-0.0.1/metropolis.css" rel="stylesheet" />
    <link href="ComputerPractice3_from_clusters_to_genes_files/remark-css-0.0.1/metropolis-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">

class: left, middle
####Computer Practice 3

##Single cell Transcriptomics using R
##Seurat: from clusters to marker genes
####Victoria Mironova 
####Associate Professor, Department of Plant Systems Physiology
&lt;img src="Figures/Theme2.png" width="100px" align = 'right'&gt;

---
class: middle, inverse


.pull-left[
#Course structure

- C1. From reads to counts
- C2. From counts to clusters
- *C3. From clusters to marker genes*
- C4. Subsetting and integrating the data.
- C5. Developmental trajectories, developmental states.

The course materials can be found via [GitHub](https://github.com/VictoriaVMironova/ScTranscriptomics_in_plants)



]
---
#What we managed doing yesterday

.pull-left[

```r
DimPlot(leaf.dataset, label = TRUE, pt.size = 1.5, label.size = 10) + NoLegend()
```

&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/DimPLot-1.png" width="504" /&gt;
]
.pull-right[

```r
DoHeatmap(leaf.dataset, features = marker_genes) + NoLegend()
```

&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/featureplot-1.png" width="504" /&gt;
]
---
class: inverse, middle

#The learning goals for this tutorial:

- how to find and visualize marker genes

- How to find differentially expressed genes?

- How to annotate the clusters?

---
#Step 1: Preparatory
Lets reuse the dataset we generated in computer practice 2:


```r
leaf.dataset&lt;-readRDS('Data/leaf.dataset.rds')
```

---
#Step 1: Preparatory
Or we can generate it once more using the summary script:


```r
path &lt;- 'Data/GSM5097888/'
leaf.dataset&lt;-Read10X(path, gene.column = 1)
leaf.dataset &lt;- CreateSeuratObject(counts = leaf.dataset, project = "leaf")
leaf.dataset[["percent.mt"]] &lt;- PercentageFeatureSet(leaf.dataset, pattern = "^ATM")
leaf.dataset[["percent.ct"]] &lt;- PercentageFeatureSet(leaf.dataset, pattern = "^ATC")
leaf.dataset &lt;- subset(leaf.dataset, subset = percent.mt &lt;= 20 &amp; percent.ct &lt;= 20 &amp; nCount_RNA &gt;=1000)
leaf.dataset &lt;- SCTransform(leaf.dataset)
leaf.dataset &lt;- RunPCA(leaf.dataset,verbose = FALSE)
leaf.dataset &lt;- RunUMAP(leaf.dataset, dims = 1:50, verbose = FALSE)
leaf.dataset &lt;- FindNeighbors(leaf.dataset, dims = 1:50, verbose = FALSE)
leaf.dataset &lt;- FindClusters(leaf.dataset, resolution = 1, verbose = FALSE)
```
*the path will be different for you

---
class: middle, inverse

#Aim 1: 

- to identify the marker genes for the cell populations

---
#Finding markers for all the clusters

FindAllMarkers is the Seurat function that finds the marker genes.


```r
leaf.markers &lt;- FindAllMarkers(leaf.dataset, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
```

--
!That is a time-consuming step, so I advice to save the results and work with the saved version further on:

```r
write.csv(leaf.markers, file = "Data/leaf_markers.csv")
```



---
#Vizualizing the markers: table

```r
leaf.markers %&gt;% 
    group_by(cluster) %&gt;%
    slice_max(n = 2, order_by = avg_log2FC)
```

```
## # A tibble: 40 x 8
## # Groups:   cluster [20]
##    X              p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene     
##    &lt;chr&gt;          &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;   &lt;int&gt; &lt;chr&gt;    
##  1 AT5G26000  0               3.31 1     0.194 0               0 AT5G26000
##  2 AT4G37990  0               3.04 0.993 0.247 0               0 AT4G37990
##  3 AT1G29920  0               2.58 1     0.508 0               1 AT1G29920
##  4 AT5G38410  0               2.53 1     0.549 0               1 AT5G38410
##  5 AT5G384201 2.45e-302       3.36 1     0.303 4.74e-298       2 AT5G38420
##  6 AT5G384301 0               3.18 0.995 0.2   0               2 AT5G38430
##  7 AT2G02100  2.35e-145       2.96 0.803 0.26  4.54e-141       3 AT2G02100
##  8 AT3G18280  1.01e-164       2.66 0.375 0.033 1.96e-160       3 AT3G18280
##  9 AT3G11930  0               4.63 0.953 0.057 0               4 AT3G11930
## 10 AT5G023801 2.43e-168       3.78 1     0.761 4.71e-164       4 AT5G02380
## # ... with 30 more rows
```
---
#Vizualizing the markers: DoHeatmap

.pull-left[

```r
top2&lt;-leaf.markers %&gt;% 
    group_by(cluster) %&gt;%
    slice_max(n = 2, order_by = avg_log2FC)
DoHeatmap(leaf.dataset, 
          features = top2$gene) + 
          NoLegend()
```
]
.pull-right[
&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/top2 markers heatmap out-1.png" width="504" /&gt;
]
---
#Vizualizing the markers: violine plot


```r
top5&lt;- leaf.markers %&gt;% 
    group_by(cluster) %&gt;%
    slice_max(n = 5, order_by = avg_log2FC)%&gt;%
    filter(cluster == 0)
VlnPlot(leaf.dataset, features = top5$gene, pt.size = 0.2, ncol = 5)
```

&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/Vln for markers-1.png" width="1008" /&gt;

---
#Vizualizing the markers: dot plot

```r
DotPlot(leaf.dataset, features = unique(top2$gene), cols = c("blue", "red"), dot.scale = 8)+ coord_flip()
```

&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/dot plot-1.png" width="864" /&gt;

---
class: middle, inverse

#Aim 2: 

- Finding differentially expressed genes

---
#What is different between two close clusters?

.pull-left[
&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/clusters adjust res 1 out2-1.png" width="504" /&gt;
]

.pull-right[
Here is a DimPlot with 19 clusters detected under default parameters.&lt;/br&gt;

We already noticed that clusters 1 and 2 are very similar in terms of the gene expression.&lt;/br&gt;

But what is the difference?

Lets find it.
]

---
#Differentially expressed genes between two clusters


```r
cluster2.markers &lt;- FindMarkers(leaf.dataset, ident.1 = 2, ident.2 = 1, min.pct = 0.25)
head(cluster2.markers, n = 8)
```

```
##                   p_val avg_log2FC pct.1 pct.2     p_val_adj
## AT1G67090 1.629544e-139  0.7676939 1.000 1.000 3.152352e-135
## AT5G54270 6.545078e-111  0.9326806 1.000 0.997 1.266145e-106
## AT5G38430 1.698622e-109  1.7572265 0.995 0.833 3.285984e-105
## AT2G38870 1.824249e-101 -1.9697493 0.739 0.978  3.529010e-97
## AT5G38420  1.006660e-97  1.0977529 1.000 0.989  1.947383e-93
## AT2G41430  4.507976e-95 -1.3376009 0.935 1.000  8.720679e-91
## AT1G07590  3.643739e-92 -1.9698584 0.862 0.989  7.048814e-88
## AT3G08940  6.881161e-90  1.0450857 0.997 0.898  1.331161e-85
```
--

- p_val : p_val (unadjusted)
- avg_log2FC : log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.
- pct.1 : The percentage of cells where the feature is detected in the first group
- pct.2 : The percentage of cells where the feature is detected in the second group
- p_val_adj : Adjusted p-value, based on Bonferroni correction using all features in the dataset.


---
#Differentially expressed genes between two clusters

```r
top10 &lt;- cluster2.markers %&gt;% slice_max(n = 10, order_by = avg_log2FC) %&gt;% row.names()
VlnPlot(leaf.dataset, features = top10, pt.size = 0.2, ncol = 5)
```

&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/Vln for markers 2 clusters-1.png" width="1152" /&gt;
The data suggests that there is no any specifically expressed gene between two clusters, they all express in both cell populations, but with different expression levels. 
---
#Other solutions for markers genes

Differentially expressed genes between a set of clusters:

```r
FindMarkers(leaf.datase, ident.1 = 2, ident.2 = c(1,6), min.pct = 0.25)
```

Seurat has several tests for differential expression which can be set with the test.use parameter.


```r
FindMarkers(leaf.dataset, ident.1 = 2, ident.2 = c(1,6), min.pct = 0.25, test.use = "roc")
```
For example, the ROC test returns the ‘classification power’ for any individual marker (ranging from 0 - random, to 1 - perfect).
Among other alternatives: “wilcox”, “bimod”, “roc”, “t”, “poisson”, “negbinom”, “LR”, “MAST”, “DESeq2”. 

---
class: middle, inverse

#Aim 3

- Annotate cell clusters

---
#Step 7. Clusters annotation by an expert

.pull-left[

```r
DimPlot(leaf.dataset, label = TRUE, pt.size = 1.5, label.size = 10) + NoLegend()
```

&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/DimPLot 3-1.png" width="504" /&gt;
]
.pull-right[

```r
DoHeatmap(leaf.dataset, features = marker_genes) + NoLegend()
```

&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/featureplot1-1.png" width="504" /&gt;
]
---
#Step 7. Clusters annotation by an expert
.pull-left[

```r
DoHeatmap(leaf.dataset, features = marker_genes) + NoLegend()
```

&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/DoHetamap clust-1.png" width="504" /&gt;
]
.pull-right[
- AT2G05100 is LHCB2, LIGHT-HARVESTING CHLOROPHYLL B-BINDING 2
- AT5G01530, AT4G10340, AT3G54890, AT3G08940 are other subunits of LIGHT HARVESTING COMPLEX PHOTOSYSTEM II
- AT1G67090, AT5G38410, AT5G38430 encode Rubisco small subunits (RBCS).
...

--

What are the clusters 1 and 2 then?

]
---
#Step 7. Clusters reannotation

When you are ready to name a cluster you can use `RenameIdents` function. Here we will rename clusters 1 and 2 into 'Mesophyll' cell type.

```r
leaf.dataset &lt;- RenameIdents(leaf.dataset,
                             `1`="Mesophyll",
                             `2`='Mesophyll')
leaf.dataset$CellTypes&lt;-Idents(leaf.dataset)
```

---
#Step 7. Clusters reannotation
.pull-left[

```r
DimPlot(leaf.dataset, label = TRUE, pt.size = 1.5, label.size = 10) + NoLegend()
```

&lt;img src="ComputerPractice3_from_clusters_to_genes_files/figure-html/DimPLot 4-1.png" width="504" /&gt;
]
.pull-right[
&lt;div class="figure"&gt;
&lt;img src="Figures/ScLeaf.jpg" alt="Graphical summary from (Lopez et al., 2021)" width="80%" /&gt;
&lt;p class="caption"&gt;Graphical summary from (Lopez et al., 2021)&lt;/p&gt;
&lt;/div&gt;

]
---
class: middle, inverse
#Your turn:

- Explore gene expression patterns for your dataset (use heatmaps, dotplots, and violine plot).

- Name a few clusters in your dataset.

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
