---
title: "scTrancriptomics in plants"
subtitle: "Tutorial 3"
## date: "24/04/2023"

output:
  xaringan::moon_reader:
    css: [default, metropolis, metropolis-fonts]
    seal: false
    nature:
      ratio: '16:9'
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---
class: left, middle
####Computer Practice 3

##Single cell Transcriptomics using R
##Seurat: from clusters to marker genes
####Victoria Mironova 
####Associate Professor, Department of Plant Systems Physiology
<img src="Figures/Theme2.png" width="100px" align = 'right'>

---
class: middle, inverse
```{r include = FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, fig.retina = 3)
set.seed(100)
```

.pull-left[
#Course structure

- C1. From reads to counts
- C2. From counts to clusters
- *C3. From clusters to marker genes*
- C4. Subsetting and integrating the data.
- C5. Developmental trajectories, developmental states.

The course materials can be found via [GitHub](https://github.com/VictoriaVMironova/ScTranscriptomics_in_plants)


```{r libraries,  echo = FALSE}
library(Seurat)
library(tidyverse)
leaf.dataset<-readRDS('Data/leaf.dataset.rds')
marker_genes<- c('AT2G05100', 'AT5G01530', 'AT4G10340', 'AT1G67090', 'AT5G38410', 'AT3G54890', 'AT1G29920', 'AT3G08940', 'AT3G27690', 'AT5G38430', 'AT1G76570', 'AT4G21750', 'AT2G26330', 'AT3G45640', 'AT2G42840', 'AT5G53210', 'AT5G60880', 'AT2G26330', 'AT1G80080', 'AT1G34245', 'AT3G26744', 'AT1G12860', 'AT2G42840', 'AT5G53210', 'AT5G60880', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT1G80080', 'AT1G34245', 'AT1G04110', 'AT4G31805', 'AT3G26744', 'AT1G12860', 'AT3G45640', 'AT3G06120', 'AT5G53210', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT1G80080', 'AT1G34245', 'AT2G20875', 'AT1G04110', 'AT4G31805', 'AT3G26744', 'AT1G12860', 'AT3G45640', 'AT3G06120', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT2G20875', 'AT1G04110', 'AT4G31805', 'AT3G26744', 'AT1G12860', 'AT1G14350', 'AT3G45640', 'AT3G24140', 'AT3G06120', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT2G20875', 'AT1G04110', 'AT3G26744', 'AT1G12860', 'AT1G14350', 'AT3G45640', 'AT3G24140', 'AT2G26330', 'AT5G62230', 'AT5G07180', 'AT1G04110', 'AT1G14350', 'AT5G46240', 'AT3G45640', 'AT3G17300', 'AT5G46240', 'AT2G26330', 'AT3G45640', 'AT4G12970', 'AT2G17950', 'AT2G27250', 'AT2G45190', 'AT2G27250', 'AT4G17710', 'AT1G55580', 'AT2G22850', 'AT3G28730', 'AT4G32880', 'AT5G16560')

```
]
---
#What we managed doing yesterday

.pull-left[
```{r DimPLot, fig.height = 6}
DimPlot(leaf.dataset, label = TRUE, pt.size = 1.5, label.size = 10) + NoLegend()
```
]
.pull-right[
```{r featureplot, fig.height = 6}
DoHeatmap(leaf.dataset, features = marker_genes) + NoLegend()
```
]
---
class: inverse, middle

#The learning goals for this tutorial:

- how to find and visualize marker genes

- How to find differentially expressed genes?

- How to annotate the clusters?

---
#Step 1: Preparatory
Lets reuse the dataset we generated in computer practice 2:

```{r dataset reuse, eval = FALSE}
leaf.dataset<-readRDS('Data/leaf.dataset.rds')
```

---
#Step 1: Preparatory
Or we can generate it once more using the summary script:

```{r dataset regenerate, eval = FALSE}
path <- 'Data/GSM5097888/'
leaf.dataset<-Read10X(path, gene.column = 1)
leaf.dataset <- CreateSeuratObject(counts = leaf.dataset, project = "leaf")
leaf.dataset[["percent.mt"]] <- PercentageFeatureSet(leaf.dataset, pattern = "^ATM")
leaf.dataset[["percent.ct"]] <- PercentageFeatureSet(leaf.dataset, pattern = "^ATC")
leaf.dataset <- subset(leaf.dataset, subset = percent.mt <= 20 & percent.ct <= 20 & nCount_RNA >=1000)
leaf.dataset <- SCTransform(leaf.dataset)
leaf.dataset <- RunPCA(leaf.dataset,verbose = FALSE)
leaf.dataset <- RunUMAP(leaf.dataset, dims = 1:50, verbose = FALSE)
leaf.dataset <- FindNeighbors(leaf.dataset, dims = 1:50, verbose = FALSE)
leaf.dataset <- FindClusters(leaf.dataset, resolution = 1, verbose = FALSE)
```
*the path will be different for you

---
class: middle, inverse

#Aim 1: 

- to identify the marker genes for the cell populations

---
#Finding markers for all the clusters

FindAllMarkers is the Seurat function that finds the marker genes.

```{r all markers, eval = FALSE}
leaf.markers <- FindAllMarkers(leaf.dataset, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
```

--
!That is a time-consuming step, so I advice to save the results and work with the saved version further on:
```{r saving markers, eval = FALSE}
write.csv(leaf.markers, file = "Data/leaf_markers.csv")
```

```{r loading markers, echo = FALSE}
leaf.markers<- read.csv(file = "Data/leaf_markers.csv", header = TRUE)
```

---
#Vizualizing the markers: table
```{r top2 markers}
leaf.markers %>% 
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
```
---
#Vizualizing the markers: DoHeatmap

.pull-left[
``` {r top2 markers heatmap, eval = FALSE}
top2<-leaf.markers %>% 
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
DoHeatmap(leaf.dataset, 
          features = top2$gene) + 
          NoLegend()
```
]
.pull-right[
``` {r top2 markers heatmap out, eval = TRUE, echo = FALSE, ref.label = "top2 markers heatmap"}
```
]
---
#Vizualizing the markers: violine plot

```{r Vln for markers, fig.height= 4.5, fig.width= 16}
top5<- leaf.markers %>% 
    group_by(cluster) %>%
    slice_max(n = 5, order_by = avg_log2FC)%>%
    filter(cluster == 0)
VlnPlot(leaf.dataset, features = top5$gene, pt.size = 0.2, ncol = 5)
```

---
#Vizualizing the markers: dot plot
```{r dot plot, fig.width= 12}

DotPlot(leaf.dataset, features = unique(top2$gene), cols = c("blue", "red"), dot.scale = 8)+ coord_flip()
```

---
class: middle, inverse

#Aim 2: 

- Finding differentially expressed genes

---
#What is different between two close clusters?

.pull-left[
```{r clusters adjust res 1 out2, echo = FALSE, eval = TRUE, ref.label='DimPLot'}
```
]

.pull-right[
Here is a DimPlot with 19 clusters detected under default parameters.</br>

We already noticed that clusters 1 and 2 are very similar in terms of the gene expression.</br>

But what is the difference?

Lets find it.
]

---
#Differentially expressed genes between two clusters

```{r find markers between two clusters}
cluster2.markers <- FindMarkers(leaf.dataset, ident.1 = 2, ident.2 = 1, min.pct = 0.25)
head(cluster2.markers, n = 8)
```
--

- p_val : p_val (unadjusted)
- avg_log2FC : log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.
- pct.1 : The percentage of cells where the feature is detected in the first group
- pct.2 : The percentage of cells where the feature is detected in the second group
- p_val_adj : Adjusted p-value, based on Bonferroni correction using all features in the dataset.


---
#Differentially expressed genes between two clusters
```{r Vln for markers 2 clusters, fig.width = 16}
top10 <- cluster2.markers %>% slice_max(n = 10, order_by = avg_log2FC) %>% row.names()
VlnPlot(leaf.dataset, features = top10, pt.size = 0.2, ncol = 5)
```
The data suggests that there is no any specifically expressed gene between two clusters, they all express in both cell populations, but with different expression levels. 
---
#Other solutions for markers genes

Differentially expressed genes between a set of clusters:
```{r markers 3 clusters, eval = FALSE}
FindMarkers(leaf.dataset, ident.1 = 2, ident.2 = c(1,6), min.pct = 0.25)
```

Seurat has several tests for differential expression which can be set with the test.use parameter.

```{r markers stat, eval = FALSE}
FindMarkers(leaf.dataset, ident.1 = 2, ident.2 = c(1,6), min.pct = 0.25, test.use = "roc")
```
For example, the ROC test returns the ‘classification power’ for any individual marker (ranging from 0 - random, to 1 - perfect).
Among other alternatives: “wilcox”, “bimod”, “roc”, “t”, “poisson”, “negbinom”, “LR”, “MAST”, “DESeq2”. 

---
class: middle, inverse

#Aim 3

- Annotate cell clusters

---
#Step 7. Clusters annotation by an expert

.pull-left[
```{r DimPLot 3, fig.height = 6}
DimPlot(leaf.dataset, label = TRUE, pt.size = 1.5, label.size = 10) + NoLegend()
```
]
.pull-right[
```{r featureplot1, fig.height = 6}
DoHeatmap(leaf.dataset, features = marker_genes) + NoLegend()
```
]
---
#Step 7. Clusters annotation by an expert
.pull-left[
```{r DoHetamap clust, fig.height = 6}
DoHeatmap(leaf.dataset, features = marker_genes) + NoLegend()
```
]
.pull-right[
- AT2G05100 is LHCB2, LIGHT-HARVESTING CHLOROPHYLL B-BINDING 2
- AT5G01530, AT4G10340, AT3G54890, AT3G08940 are other subunits of LIGHT HARVESTING COMPLEX PHOTOSYSTEM II
- AT1G67090, AT5G38410, AT5G38430 encode Rubisco small subunits (RBCS).
...

--

What are the clusters 1 and 2 then?

]
---
#Step 7. Clusters reannotation

When you are ready to name a cluster you can use `RenameIdents` function. Here we will rename clusters 1 and 2 into 'Mesophyll' cell type.
```{r clusters annotation}
leaf.dataset <- RenameIdents(leaf.dataset,
                             `1`="Mesophyll",
                             `2`='Mesophyll')
leaf.dataset$CellTypes<-Idents(leaf.dataset)
```

---
#Step 7. Clusters reannotation
.pull-left[
```{r DimPLot 4, fig.height = 6}
DimPlot(leaf.dataset, label = TRUE, pt.size = 1.5, label.size = 10) + NoLegend()
```
]
.pull-right[
```{r Ara leaf, echo=FALSE, fig.cap="Graphical summary from (Lopez et al., 2021)", out.width = '80%'}
knitr::include_graphics("Figures/ScLeaf.jpg")
```

]
---
class: middle, inverse
#Your turn:

- Explore gene expression patterns for your dataset (use heatmaps, dotplots, and violine plot).

- Name a few clusters in your dataset.

