---
title: "scTrancriptomics in plants"
subtitle: "Tutorial 5"
## date: "24/04/2023"

output:
  xaringan::moon_reader:
    css: [default, metropolis, metropolis-fonts]
    seal: false
    nature:
      ratio: '16:9'
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---

class: left, middle
####Computer Practice 5

##Single cell Transcriptomics using R
##Developmental trajectories, developmental states.
####Victoria Mironova 
####Associate Professor, Department of Plant Systems Physiology
<img src="Figures/Theme2.png" width="100px" align = 'right'>

---
class: middle, inverse
```{r include = FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, fig.retina = 3)
set.seed(100)
```

#Course structure

- C1. From reads to counts
- C2. From counts to clusters
- C3. From clusters to marker genes
- C4. Subsetting and integrating the data
- *C5. Developmental trajectories, developmental states.*

The course materials can be found via [GitHub](https://github.com/VictoriaVMironova/ScTranscriptomics_in_plants)


```{r libraries,  echo = FALSE}
library(Seurat)
library(tidyverse)
leaf.dataset<-readRDS('Data/leaf.dataset.rds')
```
---

